### PR TITLE
[ghc-api]: upgrade to ghc-9.8

### PR DESCRIPTION
**Pull request by shayne-fletcher**
update hlint to the ghc-9.8 api. fixes https://github.com/ndmitchell/hlint/issues/1545.

### CODE REVIEW

File: src/Hint/Unsafe.hs [hunk 1]
```diff
@@ -60,13 +60,15 @@ unsafeHint _ (ModuleEx (L _ m)) = \ld@(L loc d) ->
      -- 'x' is not marked 'NOINLINE'.
      , x `notElem` noinline]
   where
+    noInline = fsLit $ '{' : '-' : '#' : " NOINLINE"
```
Comment (zliu41) [inline src/Hint/Unsafe.hs:63]: Why not just `"{-# NOINLINE"`?

File: src/Hint/NumLiteral.hs [hunk 2]
```diff
@@ -49,18 +50,18 @@ numLiteralHint _ modu =
 
 suggestUnderscore :: LHsExpr GhcPs -> [Idea]
 suggestUnderscore x@(L _ (HsOverLit _ ol@(OverLit _ (HsIntegral intLit@(IL (SourceText srcTxt) _ _))))) =
-  [ suggest "Use underscore" (reLoc x) (reLoc y) [r] | '_' `notElem` srcTxt, srcTxt /= underscoredSrcTxt ]
+  [ suggest "Use underscore" (reLoc x) (reLoc y) [r] | '_' `notElem` unpackFS srcTxt, unpackFS srcTxt /= underscoredSrcTxt ]
```
Comment (zliu41) [inline src/Hint/NumLiteral.hs:53]: Make a binding (or use view pattern) for `unpackFS srcTxt`?

**Review by zliu41 — APPROVED** [review]

**Comment by shayne-fletcher** [comment]
taking into account https://github.com/ndmitchell/neil/pull/58, CI looks good https://github.com/ndmitchell/hlint/actions/runs/6519003805
Screenshot 2023-10-14 at 2 11 16 PM

**Comment by shayne-fletcher** [comment]
to enable ghc-9.8 in CI we require `cabal new-install apply-refact` succeed with ghc-9.8.1 as a build compiler which at least requires ghc-exactprint > 1.7.0.1 see https://github.com/alanz/ghc-exactprint/issues/131.

**Comment by zliu41** [comment]
Ok, I'll work on 9.8 support for apply-refact

**Review by ndmitchell — COMMENTED** [review]
Looks good to me! Shall we delay until apply-refact is out? Depends on the timeline for that.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -12,7 +12,7 @@ jobs:
       fail-fast: false
       matrix:
         os: [ubuntu-latest]
-        ghc: ['9.6', '9.4', '9.2']
+        ghc: ['9.6', '9.4']
```
[the remark this answers is not in the document]
Comment (shayne-fletcher) [reply .github/workflows/ci.yml]: we should. i've pushed that now.

**Comment by shayne-fletcher** [comment]
> Looks good to me! Shall we delay until apply-refact is out? Depends on the timeline for that.

sounds good. i think it would be reasonable to leave this here for the moment and see how the apply-refact release progresses.

**Comment by shayne-fletcher** [comment]
adding ghc-9.8 to CI requires https://github.com/ndmitchell/neil/pull/59

**Comment by shayne-fletcher** [comment]
> adding ghc-9.8 to CI requires [ndmitchell/neil#59](https://github.com/ndmitchell/neil/pull/59)

this is now landed

**Comment by shayne-fletcher** [comment]
> Ok, I'll work on 9.8 support for apply-refact

hi @zliu41 ghc-exactprint-1.8.0.0 is now on hackage it seems (https://github.com/alanz/ghc-exactprint/pull/134 & https://hackage.haskell.org/package/ghc-exactprint-1.8.0.0). ` cabal new-install apply-refact` with ghc-9.8.1 as the build compiler fails indicating there's a few modifications to be made.

**Comment by zliu41** [comment]
@shayne-fletcher Done - I just published apply-refact-0.14.0.0 on Hackage.

**Comment by shayne-fletcher** [comment]
@ndmitchell i believe this is now good to go.

**Pull request merged**